### ISSUE TITLE
Hybrid closure-proxy (acceleration)

### ISSUE THREAD

**Issue by jfparadis** [ajvincent/es7-membrane#87]
In the README, a distinction is made between membranes made of closures and proxies. What follows is summary of a third technique. It's a performance optimization.

**Recap**
Closures and proxies are on the opposite scale of performance on two aspects:
1. **Object creation**: proxies are much faster, which we can measure using a simple scenario:

```
function myClosure(raw) {	
	this.getTitle = function() {
		return raw.title;
	}
}

function myProxy(target) {	
	let proxy = new Proxy(target, myProxyHandler);
	return proxy;
}

let myProxyHandler = {};

let raw = {title: "This is A"};
new myClosure(raw); 
new myProxy(raw);

```

Metrics do show that `myProxy` is around 7x faster than `myClosure`:
[JSPerf: closure-vs-proxy-create](https://jsperf.com/closure-vs-proxy-create)

2. **Method call**: proxies are much slower, which we can also measure by adding the handler to our simple scenario:

```
let myProxyMethods = Object.create(null, {
    getTitle: {
        value: function() { 
            return this.title;
        }
    }
});

let myProxyHandler = {
     get: function(target, name, receiver) {
        let desc = Object.getOwnPropertyDescriptor(myProxyMethods, name);
        if (desc === undefined) return undefined;
	return desc.value.bind(target);
    },
};

```

Metrics do show that `myProxy` is around 700x slower than `myClosure` suggesting that the JIT compiler can't optimize the code (most likely can't bind to a reference, just like when using `with`):
[JSPerf: closure-vs-proxy-call](https://jsperf.com/closure-vs-proxy-call)

**Introducing hybrids**
Hybrids are plain objects which inherit from a proxy:
```
myHybrid = Object.create(new Proxy(target, handler));
```
The optimization is in the handler traps: closures are lazily created as properties on the plain object to shadow the proxy, and the proxy traps are invoked only once:
```
let myProxyMethods = Object.create(null, {
	getTitle: {
		value: function(raw) {return function() { 
			return raw.title;
		}}
	}
});

let myProxyHandler = {
	get: function(secret, name) {
		console.log("get",name)
		let desc = Object.getOwnPropertyDescriptor(myProxyMethods, name);
		desc.value = desc.value.call(undefined, secret.target);
		Object.defineProperty(secret.shadow, name, desc);
		return desc.value;
	},
};

function myHybrid(target) {	
	let secret = {target: target};
	let proxy = new Proxy(secret, myProxyHandler);
	let shadow = Object.create(proxy);
	secret.shadow = shadow;
	return shadow;
}
```

Metrics do show that the performance of `myHybrid` matches or slightly betters `myClosure`.
[JSPerf: closure-vs-hybrid-call](https://jsperf.com/closure-vs-hybrid-call)

The hybrids combine lazy creation and fast access times. This improvement is most noticeable compared to proxies in situations where only few methods are called repeatedly.

**Comment by ajvincent** [comment]
1. In running your closure-vs-hybrid test on a "stock" Mozilla Firefox 54.0.1, Fedora Linux 26, closure beats hybrid consistently:

closure: 141,967,519
±2.63%
fastest

hybrid:  107,500,422
±3.98%
25% slower

It's not even close.  So what platform (and results) are you seeing a benefit to this?

2. It _looks_ like you're using the proxy as a "lazy getter" for the shadow object:  when a value is available, it defines the property on the shadow so that the property doesn't have to be looked up on the proxy again.  There's a couple problems with this approach, off the top of my head: 

- what if the underlying property descriptor has a get() method instead?
- what if the underlying property value changes after you've defined it?
- what if the proxy is revoked after you've retrieved the value?  Now you have a strong reference to a value that the proxy explicitly doesn't want you to have anymore...

3. I recently created a "lazy getter" test myself, replacing a heavyweight proxy with a lightweight proxy.  (#76, final commit) The results were, surprisingly, that the light proxy offered no significant performance improvement over the heavy one.  Personally, I'm baffled by that, but the actual results are what I have to go by.

In short, I don't see how this idea really helps in the case of a _membrane_, which is to provide wrappers around whole object graphs while preserving the 1:1 identity mapping.

I'm willing to be convinced, but like I said:  I have to go by actual numbers.

3.

**Comment by davidturissini** [comment]
@ajvincent I echo what you see with Firefox. I am consistently large perf gains using the hybrid approach in Firefox.

Chrome, however, exhibits the opposite behavior. I am seeing that the hybrid approach is much faster. I'm really not sure what is happening here, but this could be related to why you were not seeing performance gains with the lazy getter approach you tried.

**Comment by jfparadis** [comment]
Guys, it looks like you are reading this wrong! 

If you measure a 25% slowdown between `hybrid` and `closure` then it's a very good thing, because `proxy`, is *3 orders of magnitude slower* than both of them:

[JSPerf: proxy-vs-hybrid-call](https://jsperf.com/proxy-vs-hybrid-call)
 
Relative slowness on call: 
```
             Closure  |  Proxy  | Hybrid
Chrome  :       1     |   700x  |   1x
Firefox :       1     |   500x  |   1.2x
```

Bottom-line, closure is the fastest, hybrid is 25% slower, and proxy over 50,000% slower (500x)!

**Comment by ajvincent** [comment]
@jfparadis ok, that's very interesting, but it still doesn't answer point 2 I raised above.  Proxies aren't just for reflecting properties of objects on first demand.  In particular, if a proxy is revoked, that's supposed to be the end of the object's lifetime.

**Comment by jfparadis** [comment]
@ajvincent: yes, this post was all about the performance of an hybrid.

Of course we can't achieve here in a few lines the same level of sophistication as what's in the repo, but I do think that this technique has its use. Imagine for example that we need a membrane for `window` or `global`, or some object that is a constant for the lifetime of the application.

In the realm, we are adding, for perfomance, a `const` declaration inside the `with` statement because the lookup on the `with` does interfere with the JIT performance optimization:

```
with (window) {
    const {Object, Array, ... } = window; 
    eval(code)
}
```

The hybrid is basically the same mechanism taken to a different context.

**Comment by ajvincent** [comment]
I am not talking about the sophisticated code in the repo.  I am talking specifically about what a Proxy, as defined in the ECMAScript 7 specification, supports.

*sigh*

When you get down to it, that's what we're really talking about:  what does a _proxy_ do versus what does the _object graph_ and the _membrane_ do.

So, what I suggest, if you are really insistent on following this path, is that you create a set of testcases illustrating that your hybrid approach can do *everything* a built-in Proxy can do, _including supporting revocation after property lookup_, and do it faster than the proxy.

I'm not trying to be a jerk, by the way:  I have believed for a long time that working code defeats theory every time.  It's simply that I'm not convinced you've covered all the important bases.  Speed isn't everything, especially with membranes.  If I were in your position, I would turn around and define an accessor descriptor on the shadow.  A getter function, while undoubtedly slower than a simple value property on the descriptor, would give you the ability to do sanity checks.

For instance, Proxy.revocable() returns both a proxy and a revoke function.  Having a revoke function (but _not_ necessarily the same one as the Proxy.revocable invocation returns, just one that calls it) is important.  You could then write another closure so that your getter checks whether a custom revoke function has been called.

```javascript
function defineFastAccessorForProperty(obj, hiddenObj, propName) {
  var isRevoked = false, proxy, revoke;
  {
    let pr = Proxy.revocable(obj, handler);
    proxy = pr.proxy;
    revoke = function() {
      isRevoked = true;
      pr.revoke();
    }
  }

  Reflect.defineProperty(obj, propName, {
    get: () => {
      if (isRevoked)
        throw new Error("Proxy has been revoked");
      return hiddenObj[propName];
    },
    // set(val), enumerable, configurable as you please
  });
}
```

**Comment by jfparadis** [comment]
Thanks @ajvincent, this is a great post. Thanks for sharing your ideas. 

I'd be thrilled if hybrids could work everywhere proxies are used. But I do agree, this needs work. Therefore, in the simplest case, hybrids could take the form of a subclass, a wrapper, etc. for very specific use cases.

Because, as you say, pragmatism needs to balance theory. Performance does threaten adoption at times. Developers look at ways to circumvent barriers like a membrane when pressured with decisions such as time-to-market.
